### PR TITLE
Feature 3 overlapping shenanigans

### PR DESCRIPTION
**Pull request by Firebrandt**
OK!
**Changes Made & Fitting into overall project **

- More or less fully implements the OVERLAP MAXIMIZATION FEATURE/USE CASE, enhancing functionality and letting us satisfy more of the project specs. Includes implementations for all relevant interfaces and their class-implementations. I refrain from listing them all for the sake of brevity, but a lot of additional functionality has been added, going from an essentially non-functional branch to a nearly-complete branch that's just awaiting final integration. Lots of methods implemented. I honestly don't remember them all.

-  Adds a couple converters to assist in the translation of my InputDialog viewModel types and regular TimetableModel Model classes. **This should make extension a little bit easier since future manipulation of those types is now, well, easier. **

-  My use case's ARCHITECTURE IS NOW CLEAN RELATIVE TO LAST PULL REQUEST, LITERALLY REFACTORED 6+ TIMES this should substantially increase maintainability. Notable changes include switching from a super flawed architecture where my View worked w/ entities pass in through initialization, that was then passed into controllers which then passed them into the interactor, to a proper Clean Architecture that makes use of dependency injection & dependency inversion to limit dependency, and that, well, doesn't break the dependency rule constantly but that instead "flows" in the proper Clean Architecture way.
    This naturally involved the creation of various interfaces to support dependency inversion/injection, and some substantial refactoring.

  -  Also, instead of passing raw entities, I make proper use of various Model and ViewModel classes to pass data between layers now.

  -  Adds basic test classes for non-trivial methods as well, helping ensure code correctness (or at least quick diagnosis of breaks). Implemented equals() methods for several model classes to support unit tests and to make writing them in future more convenient. This was also missing from the first pull request! Does not cover super large numbers of edge cases, but there could be time to add that later on.

  -  Adds substantial initialization/setup code allowing for some integration with other use cases (expect full integration to come later though). Still, increases functionality somewhat since we can move through a flow of sorts now. This was also missing from the first pull request. TThis brings us closer to full integration with a (mostly) complete application, and allows for manual testing of the use case, which wasn't possible before since the use case wasn't previously runnable.

  -  Also fixed my file structure, so it matches the pattern decided on by the group instead of shoving everything into random but convenient directories.

  -  Sidenote: Much completed branch code from earlier commits was seemingly pulled before :p. I know this PR size seems underwhelming but I've put a lot of work into the branch.

**Design Decisions I straight up don't remember all of these but here are a few:**

  -  Used static to implement the singleton design pattern for my Converter classes, since we would never reasonably want to initialize more than 1 of these classes. They're just method holders! These are (I think) an implementation of an Adapter

 -   Technically my Controller works like an instance of the Adapter design pattern.

  -   Decided to UNDO an Observer Design Pattern from an earlier commit that at one point connted my Controller to my Dialog. I, uh, kind of realized that I was being trigger happy and applying a design pattern in a situation where instead of making code more maintainable, actually makes the code more overcomplicated and harder to maintain.

   -  Implemented the Observer Design Pattern for real to listen to JD's Timetable Generating Interactor, since in that particular case I DO need an interactor to be notified precisely when he has some use-case relevant data ready to send. Implemented this via Java's built in Flow interface!

    **- Refactored original architecture quite a few times to make the architecture properly Clean.**

   - Although it's not perfect, I had the controller and interactor methods return values without actually using the returned values in the main code. This is because it makes testing a lot easier and simpler, and I wanted tests to be simpler so that they wouldn't be flaky. Javadoc has texts warning users NOT TO depend on the returned values because it likely would break our Clean Architecture.

**To whoever is reviewing this, please LMK if there is additional work or detail you would like me to elaborate on. I'm just having trouble remembering everything since I've been working on this pull request for so long.**

### CODE REVIEW

File: src/main/java/overlap_crap_fix_locations_later/application_business/CalculateSectionHoursInteractor.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
+package overlap_crap_fix_locations_later.application_business;
```
Comment (michaelhsj) [inline src/main/java/overlap_crap_fix_locations_later/application_business/CalculateSectionHoursInteractor.java]: If this is your code, could you please rename the package to something more descriptive?

File: src/main/java/overlap_crap_fix_locations_later/application_business/CalculateSectionHoursInteractor.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+
+/**
+ * A helper interactor responsible for calculating the # of hours in a section.
+ * FIXME: Not sure if this deserves its own interactor, but I feel like if we ever changed how Section time is
+ *  calculated (idk, maybe we start incorporating UofT time into the formal start), that might be a reason below... that is not
+ *  the same as how we find the overlapping timeTable in principle.
+ */
+public class CalculateSectionHoursInteractor implements SectionHoursInputBoundary {
+
```
Comment (michaelhsj) [inline src/main/java/overlap_crap_fix_locations_later/application_business/CalculateSectionHoursInteractor.java]: If this is an overriding method, I would recommend adding an @Override, just to make sure you're really overriding the boundary's methods.

File: src/main/java/overlap_crap_fix_locations_later/application_business/OverlapGeneratedTimetableRelayInteractor.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+import java.util.List;
+import java.util.concurrent.Flow;
+
+/**
+ * An interactor whose main purpose is to listen to TimetableGeneratorInteractor, and relay it's data through to
+ * the inputDialog through a presenter.
+ * <p>
+ * I am assuming that this is subscribed to something that gives me a List<Timetable>.
+ */
+public class OverlapGeneratedTimetableRelayInteractor implements Flow.Subscriber<Object> {
```
Comment (michaelhsj) [inline src/main/java/overlap_crap_fix_locations_later/application_business/OverlapGeneratedTimetableRelayInteractor.java]: Does this interactor also implement an input boundary of any type?

File: src/main/java/overlap_crap_fix_locations_later/application_business/TimeTableMatchInteractor.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+     */
+    private Double calculateTimetableOverlap(TimetableModel mainTable, TimetableModel candidate) {
+        List<CourseModel> mainCourses = mainTable.getCourses();
+        List<CourseModel> candidateCourses = candidate.getCourses();
+
+        Double totalOverlapWeightedHrs = 0.0;
+        // Although this is terribly inefficient, note that we have at most, like... 10 courses. Of 10 sections each.
+        // Section's don't store a reference to their parent course, so this is actually necessary D:
+        // Compare every section of 2 courses...
+        for (CourseModel mainCourse : mainCourses) {
```
Comment (michaelhsj) [inline src/main/java/overlap_crap_fix_locations_later/application_business/TimeTableMatchInteractor.java]: If we have time in the future, it would be good to extract a method or two so we don't have so much nesting going on in one function, but we'll live with this for now.

File: src/main/java/overlap_crap_fix_locations_later/frameworks_and_drivers/OverlapInputDialog.form [hunk 5]
```diff
@@ -0,0 +1,93 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (michaelhsj) [inline src/main/java/overlap_crap_fix_locations_later/frameworks_and_drivers/OverlapInputDialog.form]: Make sure that the Markdown file notes the prerequisites to compile this IntelliJ UI.

File: src/test/java/overlap_maximization_use_case/TestCalculateSectionHoursInteractor.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
+//package overlap_crap_fix_locations_later.tests;
```
Comment (michaelhsj) [inline src/test/java/overlap_maximization_use_case/TestCalculateSectionHoursInteractor.java]: Is there any reason why the tests here are commented out? If we can make them work before merging, that would be for the best.

File: src/main/java/overlap_crap_fix_locations_later/application_business/TimetableMatchInputBoundary.java [hunk 7]
```diff
@@ -0,0 +1,14 @@
+package overlap_crap_fix_locations_later.application_business;
+
```
Comment (michaelhsj) [inline src/main/java/overlap_crap_fix_locations_later/application_business/TimetableMatchInputBoundary.java]: This interface could use a brief explanation of what its methods do.

**Review by michaelhsj — COMMENTED** [review]
Responding to the changes made here: https://github.com/CSC207-2022F-UofT/course-project-todo-find-a-team-name/pull/86#issue-1484118368

Okay! So this is overall looking much, _much_ cleaner than before, so I'm loving the architecture. I've left some comments on the code that you can take a look at, but the long and short is as follows:

I would change the overlap package name to something more descriptive and technical. Right now it seems quite loose.

Your OverlapGeneratedTimetableRelayInteractor doesn't seem to implement any input boundaries. I'm assuming it's the case that only other interactors call it, in which case I don't think there's any problem. But if a controller ends up calling it, make sure it implements some InputBoundary.

Your TimeTableMatchInteractor has some good code, but calculateTimetableOverlap has a 5 levels of nesting, which is a little smelly. If we have some time in the future, we should plan on extracting some of the method to make it easier to read and extend. It's a priority, but the lowest on the list.

TimetableMatchInputBoundary could use a couple sentences of documentation.

The tests in TestCalculateSectionHoursInteractor are all commented out right now. Do these tests work?

File: src/main/java/overlap_crap_fix_locations_later/application_business/CalculateSectionHoursInteractor.java (hunk 1 above)
Comment (Firebrandt) [reply src/main/java/overlap_crap_fix_locations_later/application_business/CalculateSectionHoursInteractor.java]: Done! My bad, forgot to fix this earlier. But this is why we do prs!

File: src/main/java/overlap_crap_fix_locations_later/application_business/CalculateSectionHoursInteractor.java (hunk 2 above)
Comment (Firebrandt) [reply src/main/java/overlap_crap_fix_locations_later/application_business/CalculateSectionHoursInteractor.java]: oops! Fixed!

File: src/main/java/overlap_crap_fix_locations_later/application_business/OverlapGeneratedTimetableRelayInteractor.java (hunk 3 above)
Comment (Firebrandt) [reply src/main/java/overlap_crap_fix_locations_later/application_business/OverlapGeneratedTimetableRelayInteractor.java]: There's no need in this case, since essentially its only job is to listen for data from another interactor and pass it up the food chain to the UI. Note how its only methods come from implementing Flow.Subscriber<Object> - hence the name 'Relay interactor' 😇 . 

You could look at it as not really needing an Input boundary, or you could look at it as having Flow.Subscriber<Object> as its input boundary. But we probably don't need another one on top of that.

File: src/main/java/overlap_crap_fix_locations_later/application_business/TimeTableMatchInteractor.java (hunk 4 above)
Comment (Firebrandt) [reply src/main/java/overlap_crap_fix_locations_later/application_business/TimeTableMatchInteractor.java]: I'll add the TODO, because you're right about this, but the presentation is in 5 hours so

File: src/test/java/overlap_maximization_use_case/TestCalculateSectionHoursInteractor.java (hunk 6 above)
Comment (Firebrandt) [reply src/test/java/overlap_maximization_use_case/TestCalculateSectionHoursInteractor.java]: I don't know, that might have been from somebody else. But they're uncommented and ostensibly working now!

File: src/main/java/overlap_crap_fix_locations_later/application_business/TimetableMatchInputBoundary.java (hunk 7 above)
Comment (Firebrandt) [reply src/main/java/overlap_crap_fix_locations_later/application_business/TimetableMatchInputBoundary.java]: Done! Added documentation with warnings to not use the return values.

File: src/main/java/overlap_crap_fix_locations_later/frameworks_and_drivers/OverlapInputDialog.form (hunk 5 above)
Comment (Firebrandt) [reply src/main/java/overlap_crap_fix_locations_later/frameworks_and_drivers/OverlapInputDialog.form]: I added the reqs to the README since I couldn't open the XML Properly in IntelliJ.

**Review by michaelhsj — APPROVED** [review]
Looks good! See the other Feature 3 branch, everything here is clear.
